grpc/v1/webserver.proto

```diff
@@ -0,0 +1,625 @@
+
+  // Maximum number of matches: skip processing documents for a repository in
+  // a shard once we have found ShardRepoMaxMatchCount.
+  //
+  // A compound shard may contain multiple repositories. This will most often
+  // be set to 1 to find all repositories containing a result.
+  int64 shard_repo_max_match_count = 5;
+
+  // Abort the search after this much time has passed.
+  google.protobuf.Duration max_wall_time = 6;
+
+  // FlushWallTime if non-zero will stop streaming behaviour at first and
+  // instead will collate and sort results. At FlushWallTime the results will
+  // be sent and then the behaviour will revert to the normal streaming.
+  google.protobuf.Duration flush_wall_time = 7;
+
+  // Trim the number of results after collating and sorting the
+  // results
+  int64 max_doc_display_count = 8;
+
+  // If set to a number greater than zero then up to this many number
+  // of context lines will be added before and after each matched line.
+  // Note that the included context lines might contain matches and
+  // it's up to the consumer of the result to remove those lines.
+  int64 num_context_lines = 9;
+
+  // If true, ChunkMatches will be returned in each FileMatch rather than LineMatches
+  // EXPERIMENTAL: the behavior of this flag may be changed in future versions.
+  bool chunk_matches = 10;
+
+  // EXPERIMENTAL. If true, document ranks are used as additional input for
+  // sorting matches.
+  bool use_document_ranks = 11;
+
+  // EXPERIMENTAL. When UseDocumentRanks is enabled, this can be optionally set to adjust
+  // their weight in the file match score. If the value is <= 0.0, the default weight value
+  // will be used. This option is temporary and is only exposed for testing/ tuning purposes.
+  double document_ranks_weight = 12;
+
+  // Trace turns on opentracing for this request if true and if the Jaeger address was provided as
+  // a command-line flag
+  bool trace = 13;
+
+  // If set, the search results will contain debug information for scoring.
+  bool debug_score = 14;
+
+  // SpanContext is the opentracing span context, if it exists, from the zoekt client
+  map<string, string> span_context = 15;
+}
+
+message Q {
+  oneof query {
+    RawConfig raw_config = 1;
+    Regexp regexp = 2;
+    Symbol symbol = 3;
+    Language language = 4;
+    Const const = 5;
+    Repo repo = 6;
+    RepoRegexp repo_regexp = 7;
+    BranchesRepos branches_repos = 8;
+    RepoIds repo_ids = 9;
+    RepoSet repo_set = 10;
+    FileNameSet file_name_set = 11;
+    Type type = 12;
+    Substring substring = 13;
+    And and = 14;
+    Or or = 15;
+    Not not = 16;
+    Branch branch = 17;
+  }
+}
+
+// RawConfig filters repositories based on their encoded RawConfig map.
+message RawConfig {
+  enum Flags {
+    UNKNOWN = 0x00;
+    ONLY_PUBLIC = 0x01;
+    ONLY_PRIVATE = 0x02;
+    ONLY_FORKS = 0x04;
+    NO_FORKS = 0x08;
+    ONLY_ARCHIVED = 0x10;
+    NO_ARCHIVED = 0x20;
+  }
+
+  uint64 flags = 1;
```

Comment:
rather than resimulating the bitfield, I'd just use a repeated on a normal enum here. We don't have messages with a large number of RawConfig filters in them, so no need to save space over the wire. Was implemented as a bitfield in go more because that reads nicer than needing the memory savings.